### ISSUE TITLE
Ubuntu packages hets and hets-server

### ISSUE THREAD

**Issue by eugenk** [spechub/Hets#1556]
How and when are the Ubuntu packages `hets` and `hets-server` created and rolled out?
Is this a fully automated process?
Does this happen every time the master branch updates?

**Comment by cmaeder** [comment]
I do this manually, ideally only by calling https://github.com/spechub/Hets/blob/master/utils/debian/auto-package/package_trunk (or package_server) with suitable parameters (i.e. `-s wily create`).

I noticed, however, that currently the file `rev.txt` is badly overwritten by the ppa build:
https://launchpadlibrarian.net/233344982/buildlog_ubuntu-wily-amd64.hets-server_0.99r1452189354-0ubuntu1~wily_BUILDING.txt.gz

```
make[2]: git: Command not found
rm -f rev.txt
echo  >> rev.txt
```

Other packages have been created (differently) by @jelmd

**Comment by eugenk** [comment]
Ah okay. I noticed that the package `hets-server-core` is updated less frequently than `hets`.
Could we implement a script that updates and rolls out all ubuntu packages at once?

**Comment by cmaeder** [comment]
Yes, "we" could, but please minimize my work. (btw. `hets -update` still works on amd64-wily but `hets-server -update` was never implemented.)

**Comment by eugenk** [comment]
Sure, I don't suggest you to implement the script. I think it's important to know what we need to consider when creating such a script (e.g. packages that need special treatment for their creation). But if it's all in the scripts `utils/debian/auto-package/package_*`, I think we can manage it.

**Comment by cmaeder** [comment]
@sternk created this script initially (and I messed it up, copied it to `package_server` and messed it up further)

**Comment by cmaeder** [comment]
My recent additions of `-dep-suffix ""` to `ghc -M` in `UbuntuMakefile` and `ServerMakefile` will fail for older distros (like trusty up to vivid) that use ghc-7.6! https://ghc.haskell.org/trac/ghc/ticket/9287

https://github.com/spechub/Hets/commit/fb55ef11eafa2d718e848faa3fd45b0e50c18d2e

These files are used in https://github.com/spechub/Hets/blob/master/utils/debian/auto-package/debian-common/rules  (resp. debian-server/rules). Just compiling using `-O` (or `make hets-optimized`) made the ppa build machines fail due to resource exhaustion.

**Comment by jelmd** [comment]
Just for the records: Packages shown on the PPA get built by the PPA itself using the supplied debian/rules file. So wrt. PPA packaging the local job basically consists of building a) the corresponding metadata files, b) the source archive, c) signing all this stuff and finally d) uploading it to the ppa. I use my own script to do the assembly and upload (basically they do the same as the package_* files mentioned above), however, the script does hets and hets-core in one row and ensures wrt. the src tarball, that rev.txt gets created immediately and modifies the Makefile to never throw it away/re-create it. So all hets-O.N.M-a.b.c[.d]  packages definitely use the timestamp of the last commit (i.e. seconds since 1970-01-01 00:00:00) in the repo clone used for assembly for rev.txt which gets hardcoded into hets and emitted using --version. Just to make it clear: the hets-core and hets packages should have always the same version, unless a build fails on the PPA!

FWIW: `sed -re '/^\.PHONY[[:space:]]*:/,/^[[:space:]]*$/ s/[[:space:]]*rev.txt//' -i Makefile`

[OT:] The script I use also builds the deb packages locally (e.g. http://pkg.cs.ovgu.de/LNF/linux/ubuntu/14.10/ ), however, it does it in a much more efficient way, than the totally cluttered PPA/debian Wald-und-Wiesen stuff ;-) .

Last but not least: I switch on the package machinery when Till says, that a new _release_ is needed. Doing it for each commit would IMHO smell a lot and quickly chomp up all space available for the hets PPA. If one wants a "hot"/unreleased build, one can build it by itself [or download a nightly build] and do a `dpkg -i ${pkg}.deb`, for an official release 'apt-get [dist-]upgrade [$pkg]' is the right thing to do. Therefore and because of system integrity, platform compatibility 'hets[-server] -update' doesn't make IMHO any sense, should be removed.

**Comment by cmaeder** [comment]
Do you build hets with optimization switched on? (Because this required the UbuntuMakefile hack and `ghc -M` for the separate `depend` target.)

**Comment by cmaeder** [comment]
I suppose, that you cannot upload your own .deb packages. 

We could, however, ask our users to add a different source repository to their `/etc/apt/sources.list` or  `/etc/apt/sources.list.d`, but what would be the advantage? Eventually I only want a working hets binary no matter how clumsy is was compiled.

**Comment by jelmd** [comment]
*.deb packages are built the Debian way, i.e. the utils/debian/auto-package/debian-{common,server}/rules are used to build them (so basically a "make -f rules override_dh_auto_build"). If this gets the options right, everything should be fine. Anyway, right now I still use ghc 7.6, i.e. haven't tried 7.10 yet.

Yepp, the Ubuntu PPA wants to build the packages by itself, no way to upload ready-to-use packages.

Yes, if you think, downloading the *.deb package from where ever one wants to and doing a dpkg -i ... is too inconvinient, then a separate hets-nightly PPA should be setup, which than can be used to update to the most "hot" version (and hopefully to auto-delete the older ones). This is IMHO the only clean way to do it, otherwise sooner or later one  will end up with artificats, incorrect working apps and other strange problems which in turn will waste the developer's time, because they need to deal with "ghost" problems, which just do not exist in properly assembled packages ... The _use of a ppa-nightly_ for the most recent stuff seems to be a common pattern in the debian world, so I guess ppl find this more convinient/natural than some "ominous" -update switches ...

**Comment by cmaeder** [comment]
maybe still test releases could be made for vivid since support for vivid will run out this month. 
I believe it is no problem to make frequent commits to our ppa. Newer releases supersede older ones that are eventually deleted automatically. After xenial comes out (in April) vivid stuff could be deleted for good.

**Comment by jelmd** [comment]
Yepp, done. precise failed because of obscure 'out of memory' errors and I guess, that ghc got killed because it was "thinking too long". Have no clue, whether/how one can instruct PPA to use a box with sufficient RAM (whatever this is)... I triggered a rebuild, but it was probably the same poor machine ...

As long as no one vetos, I would leave it as is (would drop it in 2 month anyway).

**Comment by cmaeder** [comment]
This obscure 'out of memory'  is caused by "ghc --make" with optimization. That's the reason for
the separate `UbuntuMakefile` that first calls "ghc- M" to generate the dependency and then compiles using separate compilation.

**Comment by cmaeder** [comment]
Only  the (old) ghc version of precise has this memory problem so severe that it crashes. (Other versions take long but do not crash.)

**Comment by jelmd** [comment]
Ah ok, just wondered, because last time it went through and thus I guessed, that this is a build machine specific "physical" problem,

**Comment by cmaeder** [comment]
I'm missing Ubuntu hets packages for xenial!

**Comment by tillmo** [comment]
@jelmd, can you please take care of that?

**Comment by jelmd** [comment]
Haven't seen any xenial packages on https://launchpad.net/~hets/+archive/ubuntu/hets/+packages (e.g. aterm) - none of them are needed anymore?

**Comment by cmaeder** [comment]
many (haskell) packages need to be made for xenial, too.

**Comment by cmaeder** [comment]
at least aterm, list and hexpat. Maybe one wants glade, uni-udrawgraph and udrawgraph, too.

**Comment by jelmd** [comment]
Oh, I thought those belong to your holy kingdom, that I shouldn't touch unless explictly wished?

**Comment by cmaeder** [comment]
these are old hackage packages that hets needs and that I've only debianized. I could try to upload them.

**Comment by cmaeder** [comment]
I've uploaded the first haskell packages by re-uploading the wily versions with a modified changelog for xenial.

{{{{
sudo apt-add-repository -s "deb http://ppa.launchpad.net/hets/hets/ubuntu wily main"
sudo apt-get update
apt-get source haskell-uni-util
cd haskell-uni-util-2.3.0.1/
emacs -nw changelog   #bump version and change distro to xenial
debuild -S -sa -pgpg2
cd ..
dput ppa:hets/hets haskell-uni-util_2.3.0.1-6_source.changes
}}}

For `haskell-uni-util` I needed to adjust `debian/control`, too. (libghc-old-time-prof is a separate debian package in xenial)

Since I've created a new pgp key using `seahorse` I needed to add the option `-pgpg2` to `debuild` (this took most of my time to find out). The old `gpg` does not know my new pgp key.

**Comment by cmaeder** [comment]
I've also used `pbuilder-dist` to test some packages.

{{{
sudo apt-get install pbuilder debootstrap devscripts ubuntu-dev-tools cdbs haskell-devscripts
pbuilder-dist xenial create
pbuilder-dist xenial build haskell-uni-util_2.3.0.1-6.dsc
}}}

**Comment by jelmd** [comment]
Yepp, thanx a lot!!!

I probably don't have the time to update hets incl. old pkgs, etc. this week. Because I'm going onto vacation next week,  I'll probably postpone it to post-vacation queue, i.e. September (just in case you wonder, why nothing happens).

**Comment by jelmd** [comment]
Hmmm, the PPA crap doesn't show everything it has [wrt. vivid]. So wondering, whether it just doesn't pull in libghc-hexpat-dev libghc-aterm-dev libghc-uni-udrawgraph-dev pkgs or whether they have never been built/are really missing? If not sure, can you try to build them for vivid as well, plz?

**Comment by cmaeder** [comment]
I've removed utopic and vivid stuff assuming its end of life. Maybe I was too fast for vivid.

**Comment by jelmd** [comment]
:(, yes - needed 'til the end of this year.

**Comment by cmaeder** [comment]
I could re-upload the current versions for vivid, but is this really necessary? Who (else) still uses vivid?

**Comment by jelmd** [comment]
Yes, plz do!

Who else is in my POV actually not relevant. It is needed because as long as ontohub can't use pg 9.5+ (rails constraint IIRC), we can't upgrade to something more recent. So "'til end of the year" is more or less an optimistic estimation ...

**Comment by cmaeder** [comment]
I've uploaded most packages again. I can no longer make hets packages (due to changes). I don't know how fact++ is supposed to work (maybe the binary trusty package can simply be copied).  I can also not create hets-hollight-tools.

**Comment by cmaeder** [comment]
according to the ppa build log for hets-src the dependency ksh is missing.

**Comment by jelmd** [comment]
Thanx!

fact++: AFAIU basically the java stuff aka hets2fact++ interface - i.e. `OWL2/java/de/unibremen/**/*.java` - needs to be compiled. Everything else, i.e. the native libs containing fact++ itself && OWL2 interface to the native libs, are already included within the related jars (at least for linux ;-)). So no need to do more than compile and archive `OWL2/java/de/unibremen/**/*.java` because hets never explicitly invokes a/the `FaCT++` binary, too.

hets-hollight-tools: never tried by myself (shudder when seeing such dirty hacks and not wondering, that it breaks stuff sooner or later). But IIRC hets works w/o them as well, so IMHO just a questionable "performance" improvement for certain tasks ...

ksh: yes - see #1650

**Comment by jelmd** [comment]
New packages incl. meta pkgs are available now, however, until #1652 gets a thumb one needs to install the hets-libs pkg manually.

Before one installs the new packages (hets-{server,desktop}[-all]), one should remove the obsolete hets[-server-core]  packages.

Note that I dropped isabelle-installer, edarwin and z3 from the *-all meta packages, because they are not [yet] available for all supported ubuntu releases (isabelle-installer doesn't make any sense IMHO anyway, i.e. the user should follow isabelle install guide instead).

Last but not least: The 32bit ghc seems to have a memory problem on trusty 32bit and fails. So trusty 32bit is unsupported now [perhaps we should drop 32bit on all releases or is any hets consumer still running an old 32bit OS?].

**Comment by jelmd** [comment]
obsolete.
